### ISSUE TITLE
Help me please

### ISSUE THREAD

**Issue by suprohub** [LaurentMazare/tch-rs#955]
Can someone help me?
I want classify images for 3 categories, but its not workinhg. Loss go down but test percent go down TOO.
And i idk what i can to do for fix it. Its dont go over 1/3 (maybe becas its chosing random value).

Logs:
```
INFO  [sas::neural_network] Train class distribution:
INFO  [sas::neural_network] Class 1: 30.6%
INFO  [sas::neural_network] Class 0: 33.6%
INFO  [sas::neural_network] Class 2: 35.8%
INFO  [sas::neural_network] Test class distribution:
INFO  [sas::neural_network] Class 0: 41.8%
INFO  [sas::neural_network] Class 1: 30.4%
INFO  [sas::neural_network] Class 2: 27.8%
INFO  [sas::neural_network] Labels: 3
INFO  [sas::neural_network] Epoch  1 | Loss: 0.0365 | Test Acc: 51.56%
INFO  [sas::neural_network] Epoch  2 | Loss: 0.0333 | Test Acc: 51.56%
INFO  [sas::neural_network] Epoch  3 | Loss: 0.0314 | Test Acc: 42.19%
INFO  [sas::neural_network] Epoch  4 | Loss: 0.0295 | Test Acc: 18.75%
INFO  [sas::neural_network] Epoch  5 | Loss: 0.0282 | Test Acc: 28.12%
INFO  [sas::neural_network] Epoch  6 | Loss: 0.0267 | Test Acc: 18.75%
INFO  [sas::neural_network] Epoch  7 | Loss: 0.0261 | Test Acc: 25.00%
INFO  [sas::neural_network] Epoch  8 | Loss: 0.0252 | Test Acc: 35.94%
INFO  [sas::neural_network] Epoch  9 | Loss: 0.0242 | Test Acc: 34.38%
INFO  [sas::neural_network] Epoch 10 | Loss: 0.0233 | Test Acc: 42.19%
INFO  [sas::neural_network] Epoch 11 | Loss: 0.0228 | Test Acc: 32.81%
INFO  [sas::neural_network] Epoch 12 | Loss: 0.0221 | Test Acc: 31.25%
INFO  [sas::neural_network] Epoch 13 | Loss: 0.0212 | Test Acc: 18.75%
INFO  [sas::neural_network] Epoch 14 | Loss: 0.0206 | Test Acc: 31.25%
INFO  [sas::neural_network] Epoch 15 | Loss: 0.0199 | Test Acc: 28.12%
INFO  [sas::neural_network] Epoch 16 | Loss: 0.0195 | Test Acc: 31.25%
INFO  [sas::neural_network] Epoch 17 | Loss: 0.0190 | Test Acc: 40.62%
INFO  [sas::neural_network] Epoch 18 | Loss: 0.0187 | Test Acc: 32.81%
INFO  [sas::neural_network] Epoch 19 | Loss: 0.0179 | Test Acc: 26.56%
INFO  [sas::neural_network] Epoch 20 | Loss: 0.0178 | Test Acc: 35.94%
INFO  [sas::neural_network] Epoch 21 | Loss: 0.0173 | Test Acc: 32.81%
INFO  [sas::neural_network] Epoch 22 | Loss: 0.0171 | Test Acc: 29.69%
INFO  [sas::neural_network] Epoch 23 | Loss: 0.0170 | Test Acc: 29.69%
INFO  [sas::neural_network] Epoch 24 | Loss: 0.0166 | Test Acc: 29.69%
INFO  [sas::neural_network] Epoch 25 | Loss: 0.0161 | Test Acc: 28.12%
INFO  [sas::neural_network] Epoch 26 | Loss: 0.0160 | Test Acc: 32.81%
INFO  [sas::neural_network] Epoch 27 | Loss: 0.0158 | Test Acc: 28.12%
INFO  [sas::neural_network] Epoch 28 | Loss: 0.0149 | Test Acc: 28.12%
INFO  [sas::neural_network] Epoch 29 | Loss: 0.0150 | Test Acc: 29.69%
INFO  [sas::neural_network] Epoch 30 | Loss: 0.0145 | Test Acc: 34.38%
INFO  [sas::neural_network] Epoch 31 | Loss: 0.0143 | Test Acc: 26.56%
INFO  [sas::neural_network] Epoch 32 | Loss: 0.0141 | Test Acc: 34.38%
INFO  [sas::neural_network] Epoch 33 | Loss: 0.0136 | Test Acc: 29.69%
INFO  [sas::neural_network] Epoch 34 | Loss: 0.0131 | Test Acc: 26.56%
INFO  [sas::neural_network] Epoch 35 | Loss: 0.0131 | Test Acc: 26.56%
INFO  [sas::neural_network] Epoch 36 | Loss: 0.0125 | Test Acc: 28.12%
INFO  [sas::neural_network] Epoch 37 | Loss: 0.0123 | Test Acc: 28.12%
INFO  [sas::neural_network] Epoch 38 | Loss: 0.0125 | Test Acc: 29.69%
```

Code:
```rust
use std::{fs, path::Path};
use tch::{
    Device, Kind, Tensor,
    nn::{self, ModuleT, OptimizerConfig},
    vision::{
        dataset::{Dataset, augmentation},
        mobilenet,
    },
};

pub fn neural_main() -> Result<(), Box<dyn std::error::Error>> {
    let device = Device::cuda_if_available();
    let batch_size = 32;
    let num_epochs = 400;

    let dataset = load_dataset("data/training", "data/test")?;
    let num_classes = dataset.labels;

    log::info!("Labels: {num_classes}");

    let vs = nn::VarStore::new(device);
    let net = mobilenet::v2(&vs.root(), num_classes);

    let mut opt = nn::Adam::default().build(&vs, 1e-4)?;

    for epoch in 0..num_epochs {
        let mut train_loss = 0.0;
        let mut samples = 0;

        for (images, labels) in dataset.train_iter(batch_size) {
            let images = images.to_device(device);
            let labels = labels.to_device(device);

            let images = normalize(&images);

            let output = net.forward_t(&images, true);
            let loss = output.cross_entropy_for_logits(&labels);

            opt.backward_step(&loss);

            train_loss += loss.double_value(&[]);
            samples += images.size()[0];
        }

        let test_acc = test_model(&net, &dataset, batch_size, device);
        log::info!(
            "Epoch {:2} | Loss: {:.4} | Test Acc: {:.2}%",
            epoch + 1,
            train_loss / samples as f64,
            test_acc * 100.0
        );
    }

    Ok(())
}

fn normalize(images: &Tensor) -> Tensor {
    let mean = Tensor::from_slice(&[0.485, 0.456, 0.406])
        .to_device(images.device())
        .to_kind(Kind::Float)
        .view([3, 1, 1]);
    let std = Tensor::from_slice(&[0.229, 0.224, 0.225])
        .to_device(images.device())
        .to_kind(Kind::Float)
        .view([3, 1, 1]);

    ((images.to_kind(Kind::Float) / 255.0) - mean) / std
}

fn test_model(model: &dyn nn::ModuleT, dataset: &Dataset, batch_size: i64, device: Device) -> f64 {
    let mut total_correct = 0.0;
    let mut total_samples = 0.0;

    tch::no_grad(|| {
        for (images, labels) in dataset.test_iter(batch_size) {
            let images = normalize(&images.to_device(device));
            let labels = labels.to_device(device);

            let output = model.forward_t(&images, false);
            let preds = output.argmax(1, false);

            total_correct += preds.eq_tensor(&labels).sum(Kind::Float).double_value(&[]);
            total_samples += labels.size()[0] as f64;
        }
    });

    total_correct / total_samples
}

fn load_dataset(train_dir: &str, test_dir: &str) -> Result<Dataset, Box<dyn std::error::Error>> {
    let classes = get_classes(train_dir)?;
    let num_classes = classes.len() as i64;

    let (train_images, train_labels) = load_images_from_folder(train_dir, &classes)?;
    let (test_images, test_labels) = load_images_from_folder(test_dir, &classes)?;

    print_class_distribution(&train_labels, "Train");
    print_class_distribution(&test_labels, "Test");

    Ok(Dataset {
        train_images,
        train_labels,
        test_images,
        test_labels,
        labels: num_classes,
    })
}

fn get_classes(path: &str) -> Result<Vec<String>, Box<dyn std::error::Error>> {
    let mut classes = std::fs::read_dir(path)?
        .filter_map(|entry| {
            let entry = entry.ok()?;
            if entry.file_type().ok()?.is_dir() {
                Some(entry.file_name().into_string().ok()?)
            } else {
                None
            }
        })
        .collect::<Vec<_>>();

    classes.sort();
    Ok(classes)
}

fn load_images_from_folder(
    path: &str,
    classes: &[String],
) -> Result<(Tensor, Tensor), Box<dyn std::error::Error>> {
    let mut images = Vec::new();
    let mut labels = Vec::new();

    for (class_idx, class_name) in classes.iter().enumerate() {
        let class_path = Path::new(path).join(class_name);

        for entry in fs::read_dir(class_path)? {
            let entry = entry?;
            if entry.file_type()?.is_file() {
                let img_path = entry.path();

                let img = image::open(&img_path)?;
                let img = img.to_rgb8();
                let (height, width) = (img.height() as i64, img.width() as i64);
                let img_data = img.into_raw();

                let img_tensor = Tensor::from_slice(&img_data)
                    .reshape(&[height, width, 3]) // [height, width, channels]
                    .permute(&[2, 0, 1]); // [channels, height, width]

                images.push(img_tensor);
                labels.push(class_idx as i64);
            }
        }
    }

    Ok((Tensor::stack(&images, 0), Tensor::from_slice(&labels)))
}

fn print_class_distribution(labels: &Tensor, name: &str) {
    let labels_vec = labels.iter::<i64>().unwrap().collect::<Vec<i64>>();
    let total = labels_vec.len() as f64;
    let mut counts = std::collections::HashMap::new();

    for &label in &labels_vec {
        *counts.entry(label).or_insert(0.0) += 1.0;
    }

    log::info!("{} class distribution:", name);
    for (class, count) in counts {
        log::info!("Class {}: {:.1}%", class, 100.0 * count / total);
    }
}

```
